walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -45,89 +46,96 @@
       }
     }
 
-    Collections.sort(conflicts, TimeRange.ORDER_BY_START);
-
-    return conflicts;
+    return ImmutableList.sortedCopyOf(
+        TimeRange.ORDER_BY_START.thenComparing(TimeRange.ORDER_BY_END.reversed()), conflicts);
   }
 
-  public List<TimeRange> mergeOverlappingRanges(List<TimeRange> conflicts) {
+  private ImmutableList<TimeRange> mergeOverlappingRanges(List<TimeRange> conflicts) {
+    List<TimeRange> unnestedConflicts = removeNestedTimeRanges(conflicts);
     List<TimeRange> conflictingTimes = new ArrayList<>();
-    TimeRange first = null;
-    TimeRange second = null;
-    TimeRange combined = null;
-    boolean nested = false;
-    boolean useCombinedRange = false;
-
-    for (int i = 0; i < conflicts.size() - 1; i++) {
-      if (useCombinedRange) {
-        first = combined;
-        useCombinedRange = false;
-      } else {
-        first = conflicts.get(i);
-      }
+    int startIndex = 0;
 
-      second = conflicts.get(i + 1);
-      nested = false;
+    while (startIndex < unnestedConflicts.size()) {
+      int endIndex = findEndOfOverlap(unnestedConflicts, startIndex);
 
-      if (!first.overlaps(second)) {
-        conflictingTimes.add(first);
-      } else if (!first.contains(second)) {
-        combined = TimeRange.fromStartEnd(first.start(), second.end(), false);
-        useCombinedRange = true;
-      } else {
-        conflictingTimes.add(first);
-        nested = true;
+      TimeRange startRange = unnestedConflicts.get(startIndex);
+      TimeRange endRange = unnestedConflicts.get(endIndex - 1);
+      TimeRange overlap = TimeRange.fromStartEnd(startRange.start(), endRange.end(), false);
 
-        // Skip over adding the second TimeRange
-        i++;
-      }
+      conflictingTimes.add(overlap);
+
+      startIndex = endIndex;
+    }
+
+    return ImmutableList.copyOf(conflictingTimes);
+  }
+
+  private ImmutableList<TimeRange> removeNestedTimeRanges(List<TimeRange> conflicts) {
+    List<TimeRange> unnestedConflicts = new ArrayList<>();
+
+    if (conflicts.isEmpty()) {
+      return ImmutableList.copyOf(unnestedConflicts);
     }
 
-    if (conflicts.size() == 0 || nested) {
-      return conflictingTimes;
+    unnestedConflicts.add(conflicts.get(0));
+
+    for (TimeRange conflict : conflicts) {
+      if (!unnestedConflicts.get(unnestedConflicts.size() - 1).contains(conflict)) {
+        unnestedConflicts.add(conflict);
+      }
     }
 
-    if (conflicts.size() == 1) {
-      conflictingTimes.add(conflicts.get(0));
-    } else if (useCombinedRange) {
-      conflictingTimes.add(combined);
-    } else {
-      conflictingTimes.add(second);
+    return ImmutableList.copyOf(unnestedConflicts);
+  }
+
+  private int findEndOfOverlap(List<TimeRange> conflicts, int startIndex) {
```

Comment:
Even though this is a private method, I think it would help to add a comment since I feel the method name isn't clear by itself.